test/shopping_cart_test.rb

```diff
@@ -0,0 +1,137 @@
+      product2 = Product.new(:meat, 'chicken', 4.50, '2')
+      product3 = Product.new(:paper, 'tissue paper', 1.25, '1')
+
+      cart.add_product(product1)
+      cart.add_product(product2)
+      cart.add_product(product3)
+
+      assert_equal 13, cart.total_number_of_products
+    end
+
+    def test_is_full?
+      cart = ShoppingCart.new("King Soopers", "30items")
+
+      product1 = Product.new(:paper, 'toilet paper', 3.70, '10')
+      product2 = Product.new(:meat, 'chicken', 4.50, '2')
+      product3 = Product.new(:paper, 'tissue paper', 1.25, '1')
+      product4 = Product.new(:produce, 'apples', 0.99, '20')
+
+      cart.add_product(product1)
+      cart.add_product(product2)
+      cart.add_product(product3)
+
+      assert_equal 13, cart.total_number_of_products
+      assert_equal false, cart.is_full?
+      cart.add_product(product4)
+      assert_equal true, cart.is_full?
+    end
+
+    def test_products_by_category
+      cart = ShoppingCart.new("King Soopers", "30items")
+
+      product1 = Product.new(:paper, 'toilet paper', 3.70, '10')
+      product2 = Product.new(:meat, 'chicken', 4.50, '2')
+      product3 = Product.new(:paper, 'tissue paper', 1.25, '1')
+
+      cart.add_product(product1)
+      cart.add_product(product2)
+      cart.add_product(product3)
+
+      assert_equal [product1, product3], cart.products_by_category(:paper)
+    end
+
+    def test_percentage_occupied
+      cart = ShoppingCart.new("King Soopers", "30items")
+
+      product1 = Product.new(:paper, 'toilet paper', 3.70, '10')
+      product2 = Product.new(:meat, 'chicken', 4.50, '2')
+      product3 = Product.new(:paper, 'tissue paper', 1.25, '1')
+
+      cart.add_product(product1)
+      cart.add_product(product2)
+      cart.add_product(product3)
+
+      assert_equal 43.33, cart.percentage_occupied
+  end
+
+  def test_sort_products_by_quantity
+    cart = ShoppingCart.new("King Soopers", "30items")
+
+    product1 = Product.new(:paper, 'toilet paper', 3.70, '10')
+    product2 = Product.new(:meat, 'chicken', 4.50, '2')
+    product3 = Product.new(:paper, 'tissue paper', 1.25, '1')
+    product4 = Product.new(:produce, 'apples', 0.99, '20')
+
+    cart.add_product(product1)
+    cart.add_product(product2)
+    cart.add_product(product3)
+    cart.add_product(product4)
+
+    assert_equal [product4, product1, product2, product3],
+                 cart.sorted_products_by_quantity
+  end
+
+  def test_product_breakdown
+    cart = ShoppingCart.new("King Soopers", "30items")
+
+    product1 = Product.new(:paper, 'toilet paper', 3.70, '10')
+    product2 = Product.new(:meat, 'chicken', 4.50, '2')
+    product3 = Product.new(:paper, 'tissue paper', 1.25, '1')
+    product4 = Product.new(:produce, 'apples', 0.99, '20')
+
+    cart.add_product(product1)
+    cart.add_product(product2)
+    cart.add_product(product3)
+    cart.add_product(product4)
+    #minitest really hates hashes
+    assert_equal 3, cart.product_breakdown.length
```

Comment:
Hate is a strong word 😆 

While this test is passing, it is not giving you enough feedback to write the method in the way we were looking for.   When you need to test a hash in minitest, you can either wrap the hash in parens, or set an expected variable to hold the hash you are looking for, and then `assert_equal expected, cart.product_breakdown`